File: src/test/java/com/telegram/codex/interfaces/web/TelegramWebhookControllerTest.java
package com.telegram.codex.interfaces.web;

import com.telegram.codex.bootstrap.TelegramCodexApplication;
import com.telegram.codex.integration.telegram.application.webhook.TelegramWebhookHandler;
import com.telegram.codex.integration.telegram.domain.webhook.TelegramUpdate;
import com.telegram.codex.shared.config.AppProperties;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(
    controllers = TelegramWebhookController.class,
    properties = "logging.level.com.telegram.codex.web.TelegramWebhookController=OFF"
)
@ContextConfiguration(classes = {TelegramCodexApplication.class, TelegramWebhookControllerTest.TestConfig.class})
@Import(TelegramWebhookControllerTest.TestConfig.class)
class TelegramWebhookControllerTest {

    @TestConfiguration
    @EnableConfigurationProperties(AppProperties.class)
    static class TestConfig {
    }

    @Autowired
    private MockMvc mockMvc;

    @MockitoBean
    private TelegramWebhookHandler webhookHandler;

    @Test
    void rejectsInvalidSecret() throws Exception {
        mockMvc.perform(post("/telegram/webhook")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"update_id\":1}")
                .header("X-Telegram-Bot-Api-Secret-Token", "bad"))
            .andExpect(status().isUnauthorized())
            .andExpect(jsonPath("$.ok").value(false));
    }

    @Test
    void acceptsValidSecret() throws Exception {
        mockMvc.perform(post("/telegram/webhook")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"update_id\":1}")
                .header("X-Telegram-Bot-Api-Secret-Token", "secret"))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.ok").value(true));

        verify(webhookHandler).handle(any(TelegramUpdate.class));
    }

    @Test
    void returnsInternalServerErrorWhenHandlerFails() throws Exception {
        doThrow(new IllegalStateException("boom")).when(webhookHandler).handle(any(TelegramUpdate.class));

        mockMvc.perform(post("/telegram/webhook")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"update_id\":1}")
                .header("X-Telegram-Bot-Api-Secret-Token", "secret"))
            .andExpect(status().isInternalServerError())
            .andExpect(jsonPath("$.ok").value(false));
    }
}

File: src/test/kotlin/com/telegram/codex/interfaces/web/TelegramWebhookControllerTest.kt
package com.telegram.codex.interfaces.web

import com.telegram.codex.integration.telegram.application.TelegramWebhookHandler
import com.telegram.codex.integration.telegram.domain.TelegramUpdate
import com.telegram.codex.shared.AppProperties
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class TelegramWebhookControllerTest {
    @Test
    fun rejectsInvalidSecret() {
        val controller = TelegramWebhookController(properties(), Mockito.mock(TelegramWebhookHandler::class.java))

        val response = controller.create("bad", TelegramUpdate(1L, null))

        assertEquals(401, response.statusCode.value())
        assertEquals(false, response.body!!.ok)
    }

    @Test
    fun acceptsValidSecret() {
        val handler = Mockito.mock(TelegramWebhookHandler::class.java)
        val controller = TelegramWebhookController(properties(), handler)
        val payload = TelegramUpdate(1L, null)

        val response = controller.create("secret", payload)

        assertEquals(200, response.statusCode.value())
        assertEquals(true, response.body!!.ok)
        Mockito.verify(handler).handle(payload)
    }

    @Test
    fun returnsInternalServerErrorWhenHandlerFails() {
        val handler = Mockito.mock(TelegramWebhookHandler::class.java)
        val payload = TelegramUpdate(1L, null)
        Mockito.doThrow(IllegalStateException("boom")).`when`(handler).handle(payload)
        val controller = TelegramWebhookController(properties(), handler)

        val response = controller.create("secret", payload)

        assertEquals(500, response.statusCode.value())
        assertEquals(false, response.body!!.ok)
    }

    private fun properties(): AppProperties = AppProperties().apply {
        baseUrl = "https://example.com"
        telegramBotToken = "token"
        telegramWebhookSecret = "secret"
    }
}
